jni/src/faiss_index_service.cpp

```diff
@@ -57,6 +57,23 @@ void SetExtraParameters(knn_jni::JNIUtilInterface * jniUtil, JNIEnv *env,
 
 IndexService::IndexService(std::unique_ptr<FaissMethods> faissMethods) : faissMethods(std::move(faissMethods)) {}
 
+void IndexService::allocIndex(faiss::Index * index, size_t dim, size_t numVectors) {
+    if(auto * indexHNSWFlat = dynamic_cast<faiss::IndexHNSWFlat *>(index)) {
+        auto * indexFlatCodes = dynamic_cast<faiss::IndexFlatCodes *>(indexHNSWFlat->storage);
+        indexFlatCodes->codes.reserve(dim * numVectors * 4);
+        return;
+    }
+    if(auto * indexHNSWSQ = dynamic_cast<faiss::IndexHNSWSQ *>(index)) {
+        auto * indexFlatCodes = dynamic_cast<faiss::IndexFlatCodes *>(indexHNSWSQ->storage);
+        indexFlatCodes->codes.reserve(dim * numVectors * 2);
+        return;
+    }
+    if(auto * indexFlat = dynamic_cast<faiss::IndexFlat *>(index)) {
+        indexFlat->codes.reserve(dim * numVectors * 4);
```

Comment:
can avoid these numbers and send a parameter from Java which tells the number of bytes per dimension. because right now SQ is for fp16 only but there are other SQ which are coming up like byte vector where we want to send 1 byte per dimension.

Reply:
This can be handled in the faiss layer easily. ScalarQuantizer keeps track of the size of the vector with the specified quantization method. I can make that change.

Reply:
great. this is awesome.